docs/storage/commit_log/commit_log_based_storage_design.md

```diff
@@ -0,0 +1,354 @@
+
+The current master sequencer will create new STHs and publish them to the
+`STHs` topic, the remaining sequencers will run in a "mirror" mode to keep
+their local database state up-to-date with the master.
+
+#### Local DB storage
+
+This does not *need* to be transactional, because writes should be idempotent,
+but the implementation of the Trillian storage driver must buffer *all*,
+writes and only attempt to apply them to the DB instance when `Commit` is
+called.
+
+In the case of a partial commit failure, or crash of the signer, the next
+sequencing cycle should find that identical writes are re-attempted due to the
+signer process outlined below.
+
+#### Sequencing
+
+Assigning sequence numbers to queued leaves is implicitly performed by the
+addition of entries to the Kafka `Leaves` topic (this is termed *offset* in
+Kafka documentation).
+
+
+##### Abstract Signer process
+
+```golang
+func SignerRun() {
+  // if any of the below operations fail, just bail and retry
+
+  // read `dbSTH` (containing `treeRevision` and `sthOffset`) from local DB
+  dbSTH.treeRevision, dbSTH.sthOffset = tx.LatestSTH()
+
+  // Sanity check that the STH table has what we already know.
+  ourSTH := kafka.Read("STHs/<treeID>", dbSTH.sthOffset)
+  if ourSTH == nil {
+    glog.Errorf("should not happen - local DB has data ahead of STHs topic")
+    return
+  }
+  if ourSTH.expectedOffset != dbSTH.sthOffset {
+    glog.Errorf("should not happen - local DB committed to invalid STH from topic")
+    return
+  }
+  if ourSTH.timestamp != dbSTH.timestamp || ourSTH.tree_size != dbSTH.tree_size {
+    glog.Errorf("should not happen - local DB has different data than STHs topic")
+    return
+  }
+
+  // Look to see if anyone else has already stored data just ahead of our STH.
+  nextOffset := dbSTH.stdOffset
+  nextSTH := nil
+  for {
+      nextOffset++
+      nextSTH = kafka.Read("STHs/<treeID>", nextOffset)
+      if nextSTH == nil {
+        break
+      }
+      if nextSTH.expectedOffset != nextOffset {
+        // Someone's been writing STHs when they weren't supposed to be, skip
+        // this one until we find another which is in-sync.
+        glog.Warning("skipping invalid STH")
+        continue
+      }
+      if nextSTH.timestamp < ourSTH.timestamp || nextSTH.tree_size < ourSTH.tree_size {
+        glog.Fatal("should not happen - earlier STH with later offset")
+        return
+    }
+  }
+
+  if nextSTH == nil {
+    // We're up-to-date with the STHs topic (as of a moment ago) ...
+    if !IsMaster() {
+      // ... but we're not allowed to create fresh STHs.
+      return
+    }
+    // ... and we're the master. Move the STHs topic along to encompass any unincorporated leaves.
+    offset := dbSTH.tree_size
+    batch := kafka.Read("Leaves", offset, batchSize)
+    for b := range batch {
+      db.Put("/<treeID>/leaves/<b.offset>", b.contents)
+    }
+
+    root := UpdateMerkleTreeAndBufferNodes(batch, treeRevision+1)
```

Comment:
Might have been unclear. I meant that it should probably include the tx param to indicate this + same below.